### PR TITLE
Dockerize wheel for development environment

### LINKED ISSUE #498: Add docker support for development environment

### PR DESCRIPTION
**Pull request by akhil-gautam**
Fixes #498

### CODE REVIEW

**Comment by akhil-gautam** [comment]
@neerajdotname _a Please review.

**Comment by neerajdotname** [comment]
@akhil-gautam _a https://www.loom.com/share/c83e71ab4290446194088ea79835e1f4

**Comment by akhil-gautam** [comment]
@neerajdotname _a I have automated the `database.yml` creation. Please review.

**Comment by neerajdotname** [comment]
@akhil-gautam `bundle install --jobs=8` let's do that to speed up bundle install.

**Comment by neerajdotname** [comment]
@akhil-gautam https://www.loom.com/share/3350058894f649af9c098b1d11fac000

**Comment by akhil-gautam** [comment]
@neerajdotname Please take a look once again. I fixed the issues you mentioned.

**Comment by neerajdotname** [comment]
@akhil-gautam https://monosnap.com/file/fX8o2j4iQhQBGJLlfq4QTVSpmJl9aP

**Comment by akhil-gautam** [comment]
https://github.com/bigbinary/wheel/pull/488#issuecomment-837094006
@neerajdotname It seems, you have some cached containers and it is using them instead creating new ones.
Can you please try deleting the existing containers or try running `docker-compose up --build`?

**Review by yedhink — COMMENTED** [review]
Neeraj: I'm not exactly sure what's causing that issue. I think removing the containers and rebuilding the wheel containers as suggested by Akhil, would be a good step to check if the issue is due to polluted container build.
- Check which all containers are running `docker ps -a` and note down the container ID's to be removed.
- Remove the containers that needs to be rebuilt `docker rm $container_ids`
- Force rebuild `docker-compose up --build`

@akhil-gautam _a

How can we run the `webpack-dev-server` separately?

I tried `docker-compose run app ./bin/webpack-dev-server` in a separate shell. But still i'm not able to open up the UI in browser. In the browser console it's showing:
```
Refused to execute script from 'http://0.0.0.0:3000/packs/js/application-173708580875cf6f9716.js' because its MIME type ('text/html') is not executable, and strict MIME type checking is enabled.
```

Also I have added some comments. Please go through them and let me know if your thoughts.

**Comment by akhil-gautam** [comment]
@yedhink Did you try running `docker-compose up` without running `webpack-dev-server` separately?

**Comment by yedhink** [comment]
@akhil-gautam 

`docker-compose up` works. My main use case of handling webpack dev server separately is to apply code changes to backend and front-end, and corresponding restarts, as separate concerns. It helps with debugging too in my opinion.

But you are right. Let's take this up once priority issues are completed.

**Comment by neerajdotname** [comment]
Still not working for me. Here is the video https://www.loom.com/share/ed9440b44ae04b6e973868ccd3a36539

**Comment by yedhink** [comment]
@neerajdotname 

Please try out the following steps to see whether it fixes the issues with docker.

I'm not 100% sure whether the following steps would fix the issue, but still it's worth trying out.

Steps to re-build without removing all docker data:

- Delete the containers, if they exist. Check using `docker ps -a` and then `docker rm $container_ids_to_be_removed`.
- Delete the docker images for wheel. Run `docker images -a` to see the ID's. And then cherry pick and run `docker rmi -f $image_ids`.
- Delete the volumes of wheel. Run `docker volume ls` to see the volumes. Pick the necessary ones and do `docker volume rm -f $volume_ids`
- Then remove `sudo rm -rf ./tmp ./log` from **root of wheel project**, in dockerize-wheel branch. The dot(`.`) is important in `./tmp` path.
- Then redo `docker-compose up --build`

I had also run into some weird behaviour when running docker for wheel.

I ran the below mentioned single line command to wipe all of the docker data, which includes containers+images+volumes and then did docker-compose up --build and things started working for me in wheel.

**Warning: The following command will wipe all of docker data of all local docker projects and containers:**
```
docker rm -f $(docker ps -a -q) && docker rmi -f $(docker images -q) && docker volume rm -f $(docker volume ls -q)
```

**Comment by neerajdotname** [comment]
@yedhink Thank you so much for that. It worked. 🎉

File: README.md [hunk 1]
```diff
@@ -20,6 +20,13 @@ bundle exec rails server
 
 Visit http://localhost:3000 and login with email `oliver@example.com` and password `welcome`.
 
+## Docker for development environment
+
+1. Clone this repo by running `git clone https://github.com/bigbinary/wheel.git`
+2. `cd wheel` and then run `docker-compose up`.
+   Though the application would come up, we haven't created and seeded the database. Do `ctrl+c` and then run `docker-compose run app rails setup` to create and seed the database.
+3. From now onwards, we can just run `docker-compose up` from within the root of the `wheel` directory to bring up the application in a container.
+
```
Comment (yedhink) [inline README.md:29]: Let's mention about how to clear docker cache and get rid of weird errors. **Please do remove the `\` and `----> sentence` in the code blocks in suggestion, which i just used to to show the markdown within another markdown.**
```suggestion
# Fixing Docker Cache Issues

Sometimes, when we rebuild the container,  there are chances that the already existing cache can make the build process go awry. Thus, it's preferred to clean out the containers and rebuild from scratch in such scenarios.

Steps to re-build by removing only necessary docker data:

- Delete the containers, if they exist. Check using `docker ps -a` and then `docker rm $container_ids_to_be_removed`.
- Delete the docker images for wheel. Run `docker images -a` to see the ID's. And then cherry pick and run `docker rmi -f $image_ids`.
- Delete the volumes of wheel. Run `docker volume ls` to see the volumes. Pick the necessary ones and do `docker volume rm -f $volume_ids`
- Then remove temp files docker might've cached by `sudo rm -rf ./tmp ./log` from **root of wheel project**, in dockerize-wheel branch. The dot(`.`) is important in `./tmp` path.
- Then redo `docker-compose up --build`

If you want to try out something slightly more daring, yet effective, then run the following single line command to wipe all of the docker data, which includes containers+images+volumes and then did docker-compose up --build and things started working for me in wheel.

**Warning: The following command will wipe all of docker data of all local docker projects and containers:**

\``` -------> remove the slash (\) and this arrow and sentence
docker rm -f $(docker ps -a -q) && docker rmi -f $(docker images -q) && docker volume rm -f $(docker volume ls -q)
\``` -------> remove the slash (\) and this arrow and sentence
```

**Review by yedhink — COMMENTED** [review]
@akhil-gautam _a

I have one more suggestion. Please take a look and assign me the PR once done. I will give it a review. Let me know if you have any queries.  Thanks.

**Comment by akhil-gautam** [comment]
@neerajdotname _a Please review, I have cleaned this up.

**Pull request merged**